syntax = "proto3";

package usercart;

service UserCartService {
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc CreateAccount(CreateAccountRequest) returns (CreateAccountResponse);
  rpc AddToCart(UpdateCartRequest) returns (CartResponse);
  rpc RemoveFromCart(UpdateCartRequest) returns (CartResponse);
  rpc GetCart(UserRequest) returns (CartResponse);
  rpc Checkout(UserRequest) returns (CheckoutResponse);

  // Replication
  rpc ReplicateCart(ReplicateCartRequest) returns (Empty);
  rpc ReplicateUser(ReplicateUserRequest) returns (Empty);

  // Heartbeats and leader detection
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
}

message LoginRequest {
  string username = 1;
}

message LoginResponse {
  bool success = 1;
  int32 user_id = 2;
}

message CreateAccountRequest {
  string username = 1;
}

message CreateAccountResponse {
  bool success = 1;
  int32 user_id = 2;
}

message UserRequest {
  int32 user_id = 1;
}

message UpdateCartRequest {
  int32 user_id = 1;
  int32 itinerary_id = 2;
  int32 quantity = 3;
}

message CartItem {
  int32 itinerary_id = 1;
  int32 quantity = 2;
}

message CartResponse {
  repeated CartItem items = 1;
}

message CheckoutResponse {
  bool success = 1;
}

message ReplicateCartRequest {
  int32 user_id = 1;
  int32 itinerary_id = 2;
  int32 quantity = 3;
  bool is_deleted = 4;
}

message ReplicateUserRequest {
  int32 user_id = 1;
  string username = 2;
}

message HeartbeatRequest {
  string host = 1;
  int32 port = 2;
}

message HeartbeatResponse {
  bool is_leader = 1;
}

message Empty {}
